src/hotspot/os/windows/os_windows.cpp

```diff
@@ -3982,14 +3982,201 @@ bool   os::win32::_is_windows_server         = false;
 // including the latest one (as of this writing - Windows Server 2012 R2)
 bool   os::win32::_has_exit_bug              = true;
 
+int    os::win32::_major_version             = 0;
+int    os::win32::_minor_version             = 0;
+int    os::win32::_build_number              = 0;
+int    os::win32::_build_minor               = 0;
+
+void os::win32::initialize_windows_version() {
+  if (_major_version > 0) {
+    return; // nothing to do if already the version has already been set
+  }
+
+  LPWSTR kernel32_path = NULL;
+  LPWSTR version_info = NULL;
+  DWORD version_size = 0;
+
+  // Get the full path to \Windows\System32\kernel32.dll and use that for
+  // determining what version of Windows we're running on.
+  UINT buffer_size = GetSystemDirectoryW(NULL, 0);
```

Comment:
What is the reasoning behind making this call with `(NULL, 0)` parameters?

Can you please also provide more details on motivation for the changes you've made to this version checking code in general (after moving it from `print_windows_version`)? They're quite substantial, and I'm a little concerned that upstream won't approve of code that uses `goto`.

Reply:
I had implemented this function in a separate testing app and used my implementation here. Good catch, reverting this. Not relevant any more but... using NULL, 0 is a way of having the API tell you how much space is needed. This avoids issues with long paths when using MAX_PATH.